chaincode/src/sales/createTokenSale.ts

```diff
@@ -0,0 +1,272 @@
+    const compositeKey = TokenClass.buildTokenClassCompositeKey(tokenClassKey);
+    const tokenClass = await getObjectByKey(ctx, TokenClass, compositeKey);
+    soldTokenClasssKeys.push(compositeKey);
+
+    const quantityResults = validateTokenSaleQuantity(newSale.selling[index].quantity, tokenClass);
+    for (const validationMessage of quantityResults) {
+      chainValidationErrors.push(`Selling: ${validationMessage}`);
+    }
+  }
+
+  // All sold tokens in array must be unique
+  const uniqueSoldTokenCompositeKeys = soldTokenClasssKeys.filter((v, i, a) => a.indexOf(v) === i);
+  if(uniqueSoldTokenCompositeKeys.length != soldTokenClasssKeys.length) {
+    chainValidationErrors.push(`Length mismatch on sold token classes ${soldTokenClasssKeys.length} ` +
+        `and unique sold token classes ${uniqueSoldTokenCompositeKeys.length}. All tokens must be unique.`);
+  }
+
+  for (let index = 0; index < newSale.cost.length; index++) {
+    const tokenClassKey = newSale.cost[index].tokenInstance.getTokenClassKey();
+    const compositeKey = TokenClass.buildTokenClassCompositeKey(tokenClassKey);
+    const tokenClass = await getObjectByKey(ctx, TokenClass, compositeKey);
+
+    // Cost and sold tokens cannot be the same
+    if (soldTokenClasssKeys.includes(compositeKey)) {
+      chainValidationErrors.push(
+        `Error validating tokens to being offered: cost and purchase tokens cannot be the same.`
+      );
+    }
+
+    const quantityResults = validateTokenSaleQuantity(newSale.cost[index].quantity, tokenClass);
+    for (const validationMessage of quantityResults) {
+      chainValidationErrors.push(`Cost: ${validationMessage}`);
+    }
+  }
+
+  // All cost tokens in array must be unique
+  const uniqueCostTokenCompositeKeys = costTokenClassKeys.filter((v, i, a) => a.indexOf(v) === i);
+  if(uniqueCostTokenCompositeKeys.length != costTokenClassKeys.length) {
+    chainValidationErrors.push(`Length mismatch on cost token classes ${costTokenClassKeys.length} ` +
+        `and unique cost token classes ${uniqueCostTokenCompositeKeys.length}. All tokens must be unique.`);
+  }
+
+  // Ensure that the sale token is one that the user is an authority of 
+  // At a later date we can support users with mint allowances as well
+  // Also ensure enough supply remains to mint the token with granted allowances
+  if (chainValidationErrors.length === 0) {
+    // Check from balances
+    for (let index = 0; index < newSale.selling.length; index += 1) {
+      const tokenClassKey = newSale.selling[index].tokenClassKey;
+      const compositeKey = TokenClass.buildTokenClassCompositeKey(tokenClassKey);
+      const tokenClass = await getObjectByKey(ctx, TokenClass, compositeKey);
+
+      if(!tokenClass.authorities.includes(newSale.owner)) {
+        chainValidationErrors.push(
+          `${newSale.owner} is not an authority on ${tokenClassKey.toStringKey()}`
+        );
+      }
+
+      // TODO: Is this the correct way to check for remaining supply?
```

Comment:
tbh I'm not sure - maybe @sentientforest ?

Reply:
This has been fixed, and was backwards I believe